packages/reactor/docs/planning/Operations/index.md

```diff
@@ -0,0 +1,223 @@
+
+Unlike a typical ES approach, where Actions are generally discarded or stored in an outbox pattern, each `Action` is stored with the `Operation` that is produced from it. This allows every client to re-execute each `Action` and verify the resulting `Operation`.
+
+The lack of `previousState` is simply an optimization. As the previous state is always known, it is not necessary to store it (but perhaps could be necessary for debugging).
+
+```tsx
+/**
+ * Core Operation type that combines action data with execution metadata
+ */
+type Operation<T extends Action = Action> = {
+  /** Unique operation id */
+  id: string;
+  
+  /** Position of the operation in the history */
+  index: number;
+  
+  /** The number of operations skipped with this Operation */
+  skip: number;
+
+  /** Timestamp of when the operation was added */
+  timestampUtcMs: number;
+
+  /** The action that was executed to produce this operation */
+  action: T;
+
+  /** The resulting state after the operation is executed */
+  resultingState: string;
+
+  /** Hash of the resulting state */
+  hash: string;
+};
+```
+
+### Schema
+
+See the `[IOperationStore](../Storage/IOperationStore.md)` doc for explicit db schema.
+
+### Operation Lifecycle
+
+```mermaid
+graph TD
+    Action --> IQueue --> IJobExecutor --> Operation
+```
+
+1. Action creation - An `Action` is created by the user using the document model API.
+
+2. Queueing - Actions are queued by `(documentId, scope, branch)` to ensure proper ordering.
+
+3. Execution - Actions are passed through reducers to mutate state. Operations are created from the resulting state.
+
+4. Storage - Once applied, operations are persisted in the `IOperationStore` with atomic transactions.
+
+### Idempotency
+
+Reducers are not guaranteed to be idempotent. That is, if a reducer is called with the same input twice, it may produce different results. This is intended behavior.
+
+However, we do guarantee idempotency at the `Operation` level, with deterministic identifiers. Each operation has an `opId` derived from stable properties such as: document id, scope, branch, type, action id.
+
+Submitting the same action twice results in the same `opId`, so the store can safely ignore duplicates.
+
+The `hash` field stores the expected document state after applying the operation; reducers compute the state hash during execution and compare it to this value to detect divergence. Along with the unique `(documentId, scope, branch, index)` constraint in the storage schema, these identifiers ensure that replaying operations cannot introduce inconsistent state.
+
+### Attachments
+
+See the [Attachments doc](../Attachments/index.md) for more information on attachments. These are not stored with `Operation`s directly, but with the `Action` that consumes them.
+
+### Action Error Handling
+
+Operations cannot be in a failure state, as they represent facts about the document state. Instead, the `Action` that produced the `Operation` may fail.
```

Comment:
Currently, we do have a `error` field on operations. It might happen when shuffling an operation that becomes invalid when applied to a new state.
I think the error gets bubbled up to the sync protocol as mentioned, which interrupts syncing between two Reactors.
Maybe we should instead add the shuffled operations with the error state, and the other Reactor should then deal with that error. (IE, notify the user that is action failed)